pkgs/build-support/docker/default.nix

```diff
@@ -0,0 +1,354 @@
+        rmdir --ignore-fail-on-non-empty mnt/dev mnt/proc mnt/sys mnt/nix/store mnt/nix
+      '';
+ 
+      postUmount = ''
+        pushd layer
+        ${extraCommands}
+        popd
+
+        echo Packing layer
+        mkdir $out
+        tar -C layer -cf $out/layer.tar .
+        ts=$(${tarsum} < $out/layer.tar)
+        cat ${baseJson} | jshon -s "$ts" -i checksum > $out/json
+       echo -n "1.0" > $out/VERSION
+      '';
+    };
+
+  # 1. extract the base image
+  # 2. create the layer
+  # 3. add layer deps to the layer itself, diffing with the base image
+  # 4. compute the layer id
+  # 5. put the layer in the image
+  # 6. repack the image
+  buildImage = args@{ name, tag ? "latest"
+               , fromImage ? null, fromImageName ? null, fromImageTag ? null
+               , contents ? null, tarballs ? [], config ? null
+               , runAsRoot ? null, diskSize ? 1024, extraCommands ? "" }:
+
+    let
+
+      baseJson = writeText "${name}-config.json" (builtins.toJSON {
+          created = "1970-01-01T00:00:01Z";
+          architecture = "amd64";
+          os = "linux";
+          config = config;
+      });
+      
+      layer = (if runAsRoot == null
+               then mkPureLayer { inherit baseJson contents extraCommands; }
+               else mkRootLayer { inherit baseJson fromImage fromImageName fromImageTag contents runAsRoot diskSize extraCommands; });
+      depsTarball = mkTarball { name = "${name}-deps";
+                                drv = layer;
+                                onlyDeps = true; };
+      
+      result = runCommand "${name}.tar.gz" {
+        buildInputs = [ jshon ];
+
+        imageName = name;
+        imageTag = tag;
+        inherit fromImage baseJson;
+
+        mergedTarball = if tarballs == [] then depsTarball else mergeTarballs ([ depsTarball ] ++ tarballs);
+
+        passthru = {
+          buildArgs = args;
+        };
+      } ''
+        mkdir image
+        touch baseFiles
+        if [ -n "$fromImage" ]; then
+          echo Unpacking base image
+          tar -C image -xpf "$fromImage"
+          
+          if [ -z "$fromImageName" ]; then
+            fromImageName=$(jshon -k < image/repositories|head -n1)
+          fi
+          if [ -z "$fromImageTag" ]; then
+            fromImageTag=$(jshon -e $fromImageName -k < image/repositories|head -n1)
+          fi
+          parentID=$(jshon -e $fromImageName -e $fromImageTag -u < image/repositories)
+          
+          for l in image/*/layer.tar; do
+            tar -tf $l >> baseFiles
+          done
+        fi
+
+        chmod -R a+rw image
+        
+        mkdir temp
+        cp ${layer}/* temp/
+        chmod +w temp/*
+        
+        echo Adding dependencies
+        tar -tf temp/layer.tar >> baseFiles
+        tar -tf "$mergedTarball" | grep -v ${layer} > layerFiles
+        sed -i -e 's|^[\./]\+||' baseFiles layerFiles
+        comm <(sort -n baseFiles|uniq) <(sort -n layerFiles|uniq) -1 -3 > newFiles
```

Comment:
I'm on straight-up NixOS. Here's a small repro of what I see

```
[anders@gurney:~/playground/nix-docker]
$ nix-build gcc.nix -A image
these derivations will be built:
  /nix/store/hdchj8q53khs4lz88c219hflgghnzxil-gcc.tar.gz.drv
building path(s) ‘/nix/store/cls0x1v8g6d5zivp00cqxiddiyx9nsi9-gcc.tar.gz’
Adding dependencies
comm: file 1 is not in sorted order
builder for ‘/nix/store/hdchj8q53khs4lz88c219hflgghnzxil-gcc.tar.gz.drv’ failed with exit code 1
error: build of ‘/nix/store/hdchj8q53khs4lz88c219hflgghnzxil-gcc.tar.gz.drv’ failed
[anders@gurney:~/playground/nix-docker]
$ cat gcc.nix
with import <nixpkgs> { };
with dockerTools;

rec {
  image = buildImage {
    name = "gcc";
    contents = [ gcc distcc ];
    config = {
      Cmd = [ "${gcc}/bin/gcc" ];
    };
  };
}
```

I'm on nixpkgs 07ab8c6c01a2ca74baf5fde2fad45833ccf09bba, with this change cherry-picked on top